File: Lab 13/main.cpp
#include <iostream>
#include <stdio.h>

struct point
{
    int x;
    int y;
};
struct point my_point = {3, 7};

// binary tree

struct Node
{
    const char *name;
    Node *left; // recursion
    Node *right;
};

void traceByTree(Node node);

int main()
{
    // 1st
    struct point *p = &my_point;
    (*p).x = 8;
    p->y = 9;

    printf(" x = %d , y= %d", my_point.x, my_point.y);

    // 2nd

    Node n1, n2, n3, n4, n5;
    n1.name = "n1";
    n2.name = "n2";
    n3.name = "n3";
    n4.name = "n4";
    n5.name = "n5";

    n1.left = &n2;
    n1.right = &n3;

    n3.right = NULL;
    n3.left = NULL;

    n2.left = &n4;
    n2.right = &n5;

    n4.left = NULL;
    n4.right = NULL;

    n5 = n4;

    traceByTree(n1);
    return 0;
}

void traceByTree(Node node)
{
    printf(" node .name = %s\n", node.name);
    if (node.left != NULL)
    {
        printf("node.left \n");
        traceByTree(*node.left);
    }
    if (node.right != NULL)
    {
        printf(" node.right \n");
        traceByTree(*node.right);
    }
}

File: Lab 13/main.py
# build a binary tree

class Node:
    def __init__(self,name):
        self.name = name
        self.left = None
        self.right =None
    
    def PrintTree(self):
        print(self.name)

root = Node(0)

root.PrintTree

root.left = Node(1)
root.right = Node(2)


class Student:
    def __dir__(self) -> Iterable[str]:
        pass
    def __init__(self, name, group, test, final = 0, grade = 0):
        self.name =name
        self.group = group
        self.test =test
        self.final = final
        self.grade = grade

    def myfunc(self):
        print("hello my name is "+self.name)

Lisa = Student ('Lisa', 'A', [5,4,5])
print(Lisa.name,Lisa.group, " test 1 = ", Lisa.test[0])
Lisa.myfunc()

Bart = Student ('Bart', 'A', [4,4,5])
print(Bart.name,Bart.group, " test 1 = ", Bart.test[0])
Bart.myfunc()

Homer = Student ('Homer', 'A', [3,4,4])
print(Homer.name,Homer.group, " test 1 = ", Homer.test[0])
Homer.myfunc()

class Card:
    def __init__(self,pips,suit):
        self.pips =pips
        self.suit =suit
# Build a deck of cars
deck = []
deck.append(Card(0,"Hearts"))
deck.append(Card(1,"Hearts"))
deck.append(Card(2,"Hearts"))

for obj in deck:
    print(obj.pips,obj.suit)


